### ISSUE TITLE
AttributeError: 'DatabaseTable' object has no attribute 'table'

### ISSUE THREAD

**Issue by jhlch** [cloudera/ibis#846]
I'm using Ibis 0.8.1:

```
Sub-process traceback:
---------------------------------------------------------------------------
AttributeError                                     Wed Jun  1 14:08:29 2016
PID: 11000       Python 2.7.6: /home/juliet/.virtualenvs/release/bin/python
...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/joblib/parallel.py in __call__(self=<joblib.parallel.BatchedCalls object>)
     67     def __init__(self, iterator_slice):
     68         self.items = list(iterator_slice)
     69         self._size = len(self.items)
     70 
     71     def __call__(self):
---> 72         return [func(*args, **kwargs) for func, args, kwargs in self.items]
        func = <function process_dir>
        args = ('/home/juliet/relqual-results/as-of-2014-09-14',)
        kwargs = {}
        self.items = [(<function process_dir>, ('/home/juliet/relqual-results/as-of-2014-09-14',), {})]
     73 
     74     def __len__(self):
     75         return self._size
     76 

...........................................................................
/home/juliet/relqual-results/as-of-2014-09-14/releasequality/FitPastModels.py in process_dir(filename='/home/juliet/relqual-results/as-of-2014-09-14')
     46 
     47 
     48 
     49 
     50 
---> 51 
     52 
     53 
     54 
     55 

...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/releasequality-0.1-py2.7.egg/releasequality/RunAnalysis.py in run()
     26     unique_event_type = config.get('DataIO', 'UniqueEventType')
     27     eng_esc = config.getboolean('AndersenGill', 'EngEsc')
     28     priorities = config.get('AndersenGill', 'Priorities').split(",")
     29 
     30     start_dataio_time = time.clock()
---> 31     DataIO.main(only_est_install, version, filter_date, unique_event_type)
     32     end_dataio_time = time.clock()
     33     log_elapased_time("DataIO", start_dataio_time, end_dataio_time)
     34 
     35     start_AG_time = time.clock()

...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/releasequality-0.1-py2.7.egg/releasequality/DataIO.py in main(only_est_install=False, version_granularity='minor', filter_date_str='2014-09-14', unique_event_type='support')
    510                                               version_granularity,
    511                                               filter_date, unique_event_type),
    512                                           version_install_interval(
    513                                               cluster_upgrade_tbl,
    514                                               version_granularity,
--> 515                                               filter_date
    516                                           ),
    517         )
    518         print "Writing Results."
    519         cdh_esc_upgrade_hist_df = ibis_conn.execute(

...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/releasequality-0.1-py2.7.egg/releasequality/DataIO.py in version_install_interval(cluster_upgrade=DatabaseTable[table]
  name: u_aimam.`clustersta...n_time : string
    last_collection_time : string, version_granularity='minor', filter_date='2014-09-14')
    132     upgrade_hist = not_empty_either[diag_minor_cer_col, not_empty_either]
    133 
    134     new_cols = upgrade_hist.mutate(global_name=upgrade_hist.globalname,
    135                                    salesforce_id=upgrade_hist.salesforceid,
    136                                    diag_cluster_id=upgrade_hist.clusterid,
--> 137                                    cluster_name=upgrade_hist.clustername)
    138     # Filter out data before performing time interval estimation
    139     # actually how does
    140     if filter_date:
    141         new_cols = new_cols[new_cols]

...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/ibis/expr/api.py in mutate(table=ref_0
DatabaseTable[table]
  name: u_aimam.`clus...])
      Literal[int8]
        1
    Table: ref_2, exprs=[ref_0
DatabaseTable[table]
  name: u_aimam.`clus...umn[array(string)] 'clustername' from table ref_3, ref_0
DatabaseTable[table]
  name: u_aimam.`clus...olumn[array(string)] 'clusterid' from table ref_3, ref_0
DatabaseTable[table]
  name: u_aimam.`clus...lumn[array(string)] 'globalname' from table ref_3, ref_0
DatabaseTable[table]
  name: u_aimam.`clus...mn[array(string)] 'salesforceid' from table ref_3], **kwds={'cluster_name': ref_0
DatabaseTable[table]
  name: u_aimam.`clus...umn[array(string)] 'clustername' from table ref_3, 'diag_cluster_id': ref_0
DatabaseTable[table]
  name: u_aimam.`clus...olumn[array(string)] 'clusterid' from table ref_3, 'global_name': ref_0
DatabaseTable[table]
  name: u_aimam.`clus...lumn[array(string)] 'globalname' from table ref_3, 'salesforce_id': ref_0
DatabaseTable[table]
  name: u_aimam.`clus...mn[array(string)] 'salesforceid' from table ref_3})
   2027             if x.get_name() not in used:
   2028                 proj_exprs.append(x)
   2029 
   2030         return table.projection(proj_exprs)
   2031     else:
-> 2032         return table.projection([table] + exprs)
        table.projection = <bound method TableExpr.projection of ref_0
Data...)
      Literal[int8]
        1
    Table: ref_2>
        table = ref_0
DatabaseTable[table]
  name: u_aimam.`clus...])
      Literal[int8]
        1
    Table: ref_2
        exprs = [ref_0
DatabaseTable[table]
  name: u_aimam.`clus...umn[array(string)] 'clustername' from table ref_3, ref_0
DatabaseTable[table]
  name: u_aimam.`clus...olumn[array(string)] 'clusterid' from table ref_3, ref_0
DatabaseTable[table]
  name: u_aimam.`clus...lumn[array(string)] 'globalname' from table ref_3, ref_0
DatabaseTable[table]
  name: u_aimam.`clus...mn[array(string)] 'salesforceid' from table ref_3]
   2033 
   2034 
   2035 def projection(table, exprs):
   2036     """

...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/ibis/expr/api.py in projection(table=ref_0
DatabaseTable[table]
  name: u_aimam.`clus...])
      Literal[int8]
        1
    Table: ref_2, exprs=[ref_0
DatabaseTable[table]
  name: u_aimam.`clus...])
      Literal[int8]
        1
    Table: ref_2, ref_0
DatabaseTable[table]
  name: u_aimam.`clus...umn[array(string)] 'clustername' from table ref_3, ref_0
DatabaseTable[table]
  name: u_aimam.`clus...olumn[array(string)] 'clusterid' from table ref_3, ref_0
DatabaseTable[table]
  name: u_aimam.`clus...lumn[array(string)] 'globalname' from table ref_3, ref_0
DatabaseTable[table]
  name: u_aimam.`clus...mn[array(string)] 'salesforceid' from table ref_3])
   2051     if isinstance(exprs, (Expr,) + six.string_types):
   2052         exprs = [exprs]
   2053 
   2054     projector = L.Projector(table, exprs)
   2055 
-> 2056     op = projector.get_result()
        op = undefined
        projector.get_result = <bound method Projector.get_result of <ibis.expr.analysis.Projector object>>
   2057     return TableExpr(op)
   2058 
   2059 
   2060 def _table_relabel(table, substitutions, replacements=None):

...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/ibis/expr/analysis.py in get_result(self=<ibis.expr.analysis.Projector object>)
    680 
    681     def get_result(self):
    682         roots = self.parent_roots
    683 
    684         if len(roots) == 1 and isinstance(roots[0], ops.Selection):
--> 685             fused_op = self._check_fusion(roots[0])
        fused_op = undefined
        self._check_fusion = <bound method Projector._check_fusion of <ibis.expr.analysis.Projector object>>
        roots = [Selection(ref_0
DatabaseTable[table]
  name: u_a..._1\n        Literal[int8]\n          0"], [], [])]
    686             if fused_op is not None:
    687                 return fused_op
    688 
    689         return ops.Selection(self.parent, self.clean_exprs)

...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/ibis/expr/analysis.py in _check_fusion(self=<ibis.expr.analysis.Projector object>, root=Selection(ref_0
DatabaseTable[table]
  name: u_a..._1\n        Literal[int8]\n          0"], [], []))
    720                     fused_exprs.append(y)
    721 
    722                 # This was a filter, so implicitly a select *
    723                 if not have_root and len(root.selections) == 0:
    724                     fused_exprs = [root.table] + fused_exprs
--> 725             elif validator.validate(lifted_val):
        validator.validate = <bound method ExprValidator.validate of <ibis.expr.analysis.ExprValidator object>>
        lifted_val = ref_0
DatabaseTable[table]
  name: u_aimam.`clus...umn[array(string)] 'clustername' from table ref_3
    726                 can_fuse = True
    727                 fused_exprs.append(lifted_val)
    728             elif not validator.validate(val):
    729                 can_fuse = False

...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/ibis/expr/analysis.py in validate(self=<ibis.expr.analysis.ExprValidator object>, expr=ref_0
DatabaseTable[table]
  name: u_aimam.`clus...umn[array(string)] 'clustername' from table ref_3)
    759         return self.validate(expr)
    760 
    761     def validate(self, expr):
    762         op = expr.op()
    763         if isinstance(op, ops.TableColumn):
--> 764             if self._among_roots(op.table.op()):
        self._among_roots = <bound method ExprValidator._among_roots of <ibis.expr.analysis.ExprValidator object>>
        op.table.op = <bound method TableExpr.op of ref_0
DatabaseTabl...)
      Literal[int8]
        1
    Table: ref_2>
    765                 return True
    766         elif isinstance(op, ops.Selection):
    767             if self._among_roots(op):
    768                 return True

...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/ibis/expr/analysis.py in _among_roots(self=<ibis.expr.analysis.ExprValidator object>, node=Selection(ref_0
DatabaseTable[table]
  name: u_a..._1\n        Literal[int8]\n          0"], [], []))
    772             if not self._among_roots(root):
    773                 return False
    774         return True
    775 
    776     def _among_roots(self, node):
--> 777         return self.roots_shared(node) > 0
        self.roots_shared = <bound method ExprValidator.roots_shared of <ibis.expr.analysis.ExprValidator object>>
        node = Selection(ref_0
DatabaseTable[table]
  name: u_a..._1\n        Literal[int8]\n          0"], [], [])
    778 
    779     def roots_shared(self, node):
    780         count = 0
    781         for root in self.roots:

...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/ibis/expr/analysis.py in roots_shared(self=<ibis.expr.analysis.ExprValidator object>, node=Selection(ref_0
DatabaseTable[table]
  name: u_a..._1\n        Literal[int8]\n          0"], [], []))
    777         return self.roots_shared(node) > 0
    778 
    779     def roots_shared(self, node):
    780         count = 0
    781         for root in self.roots:
--> 782             if root.is_ancestor(node):
        root.is_ancestor = <bound method Selection.is_ancestor of Selection...m table ref_1\n    Literal[int8]\n      0"], [])>
        node = Selection(ref_0
DatabaseTable[table]
  name: u_a..._1\n        Literal[int8]\n          0"], [], [])
    783                 count += 1
    784         return count
    785 
    786     def shares_some_roots(self, expr):

...........................................................................
/home/juliet/.virtualenvs/release/local/lib/python2.7/site-packages/ibis/expr/operations.py in is_ancestor(self=Selection(ref_0
DatabaseTable[table]
  name: u_a...om table ref_1\n    Literal[int8]\n      0"], []), other=Selection(ref_0
DatabaseTable[table]
  name: u_a..._1\n        Literal[int8]\n          0"], [], []))
   1736 
   1737         table = self.table
   1738         exist_layers = False
   1739         op = table.op()
   1740         while not (op.blocks() or isinstance(op, Join)):
-> 1741             table = table.op().table
        table = DatabaseTable[table]
  name: u_aimam.`clustersta...n_time : string
    last_collection_time : string
        table.op.table = undefined
   1742             exist_layers = True
   1743 
   1744         if exist_layers:
   1745             reboxed = Selection(table, self.selections)

AttributeError: 'DatabaseTable' object has no attribute 'table'
```

**Comment by tdhopper** [comment]
I've seen this in a number of places too. I haven't been able to create a minimum viable example. Here's my current error:

```
/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/types.pyc in __getitem__(self, what)
    581         if isinstance(what, (list, tuple, TableExpr)):
    582             # Projection case
--> 583             return self.projection(what)
    584         elif isinstance(what, BooleanArray):
    585             # Boolean predicate

/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/api.pyc in projection(table, exprs)
   2054     projector = L.Projector(table, exprs)
   2055 
-> 2056     op = projector.get_result()
   2057     return TableExpr(op)
   2058 

/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/analysis.pyc in get_result(self)
    687                 return fused_op
    688 
--> 689         return ops.Selection(self.parent, self.clean_exprs)
    690 
    691     def _check_fusion(self, root):

/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/operations.pyc in __init__(self, table_expr, proj_exprs, predicates, sort_keys)
   1671 
   1672         dependent_exprs = clean_exprs + self.sort_keys
-> 1673         self._validate(dependent_exprs)
   1674         self._validate_predicates()
   1675 

/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/operations.pyc in _validate(self, exprs)
   1692         # input table; this means they must either be scalar expressions or
   1693         # array expressions originating from the same root table expression
-> 1694         self.table._assert_valid(exprs)
   1695 
   1696     def _get_schema(self, proj_exprs):

/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/types.pyc in _assert_valid(self, exprs)
    550     def _assert_valid(self, exprs):
    551         from ibis.expr.analysis import ExprValidator
--> 552         ExprValidator([self]).validate_all(exprs)
    553 
    554     def __contains__(self, name):

/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/analysis.pyc in validate_all(self, exprs)
    803     def validate_all(self, exprs):
    804         for expr in exprs:
--> 805             self.assert_valid(expr)
    806 
    807     def assert_valid(self, expr):

/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/analysis.pyc in assert_valid(self, expr)
    806 
    807     def assert_valid(self, expr):
--> 808         if not self.validate(expr):
    809             msg = self._error_message(expr)
    810             raise RelationError(msg)

/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/analysis.pyc in validate(self, expr)
    762         op = expr.op()
    763         if isinstance(op, ops.TableColumn):
--> 764             if self._among_roots(op.table.op()):
    765                 return True
    766         elif isinstance(op, ops.Selection):

/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/analysis.pyc in _among_roots(self, node)
    775 
    776     def _among_roots(self, node):
--> 777         return self.roots_shared(node) > 0
    778 
    779     def roots_shared(self, node):

/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/analysis.pyc in roots_shared(self, node)
    780         count = 0
    781         for root in self.roots:
--> 782             if root.is_ancestor(node):
    783                 count += 1
    784         return count

/Users/tdhopper/miniconda/envs/DS-env/lib/python2.7/site-packages/ibis/expr/operations.pyc in is_ancestor(self, other)
   1739         op = table.op()
   1740         while not (op.blocks() or isinstance(op, Join)):
-> 1741             table = table.op().table
   1742             exist_layers = True
   1743 

AttributeError: 'DatabaseTable' object has no attribute 'table'
```

**Comment by wesm** [comment]
This helps -- Tim, is there any way to share a sanitized version of the code where this error occurs (privately / off github is OK)? I can try to come up with a standalone breaking example.

**Comment by tdhopper** [comment]
Done

**Comment by tdhopper** [comment]
@wesm Have you figured out anything on this? I'm seeing frequently.

**Comment by cpcloud** [comment]
@tdhopper can you post the repro of this here? or a paste? or email works too.

**Comment by wesm** [comment]
I'm looking at this to see if I can concoct a failing test case

**Comment by tdhopper** [comment]
@cpcloud: I've only been able to reproduce with fairly complex examples on my work impala cluster. I sent Wes a notebook that can't be shared publicly but I can show you if you want. I can't seem to make it happen on a smaller scale.

**Comment by pteehan** [comment]
I just hit this issue.  One interesting thing is that it occurred in a notebook which was working yesterday, so the code was unchanged.  I believe the Impala server was restarted overnight.  I re-ran the whole notebook, creating the connection etc, and was still getting the error.  But when I restarted the kernel and ran it again, it worked fine.  Hope that's a useful clue.

**Comment by cpcloud** [comment]
This should be fixed by #901. @tdhopper can you give master a try and see if you can repro? pls reopen if you can.
